src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -1116,29 +1130,50 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 );
                 eq
             })
-            .map(|ty::GeneratorInteriorTypeCause { span, scope_span, expr, .. }| {
-                (span, source_map.span_to_snippet(*span), scope_span, expr)
+            .map(|cause| {
+                // Check to see if any awaited expressions have the target type.
+                let from_awaited_ty = visitor
+                    .awaits
+                    .into_iter()
+                    .map(|id| self.tcx.hir().expect_expr(id))
+                    .find(|expr| {
+                        let ty = tables.expr_ty_adjusted(&expr);
+                        // Compare types using the same logic as above.
+                        let ty_erased = self.tcx.erase_late_bound_regions(&ty::Binder::bind(ty));
+                        let ty_erased = self.tcx.erase_regions(&ty_erased);
+                        let eq = ty::TyS::same_type(ty_erased, target_ty_erased);
+                        debug!(
+                            "maybe_note_obligation_cause_for_async_await: await_expr={:?} \
+                            await_ty_erased={:?}  target_ty_erased={:?} eq={:?}",
+                            expr, ty_erased, target_ty_erased, eq
+                        );
+                        eq
```

Comment:
Would be great to avoid copy paste programming here.

Reply:
Also, the overall function is pretty huge now -- it would be good to refactor it.